source/http/HttpConnection.cpp

```diff
@@ -0,0 +1,363 @@
+                options.user_data = connectionWrapper;
+                options.host_name = hostName;
+                options.port = port;
+                options.initial_window_size = m_initialWindowSize;
+                options.socket_options = &const_cast<Io::SocketOptions &>(socketOptions);
+                options.on_setup = HttpClient::s_onClientConnectionSetup;
+                options.on_shutdown = HttpClient::s_onClientConnectionShutdown;
+
+                if (aws_http_client_connect(&options))
+                {
+                    connectionWrapper->~ConnectionWrapper();
+                    aws_mem_release(m_allocator, connectionWrapperMemory);
+                    return false;
+                }
+
+                return true;
+            }
+
+            bool HttpClient::NewConnection(
+                const ByteCursor &hostName,
+                uint16_t port,
+                const Io::SocketOptions &socketOptions) const noexcept
+            {
+                void *connectionWrapperMemory = aws_mem_acquire(m_allocator, sizeof(ConnectionWrapper));
+                if (!connectionWrapperMemory)
+                {
+                    return false;
+                }
+
+                ConnectionWrapper *connectionWrapper = new (connectionWrapperMemory) ConnectionWrapper(this);
+
+                aws_http_client_connection_options options;
+                AWS_ZERO_STRUCT(options);
+                options.self_size = sizeof(aws_http_client_connection_options);
+                options.bootstrap = m_bootstrap->GetUnderlyingHandle();
+                options.tls_options = nullptr;
+                options.allocator = m_allocator;
+                options.user_data = connectionWrapper;
+                options.host_name = hostName;
+                options.port = port;
+                options.initial_window_size = m_initialWindowSize;
+                options.socket_options = &const_cast<Io::SocketOptions &>(socketOptions);
+                options.on_setup = HttpClient::s_onClientConnectionSetup;
+                options.on_shutdown = HttpClient::s_onClientConnectionShutdown;
+
+                if (aws_http_client_connect(&options))
+                {
+                    connectionWrapper->~ConnectionWrapper();
+                    aws_mem_release(m_allocator, connectionWrapperMemory);
+                    return false;
+                }
+
+                return true;
+            }
+
+            HttpConnection::HttpConnection(aws_http_connection *connection, Allocator *allocator) noexcept
+                : m_connection(connection), m_allocator(allocator), m_lastError(AWS_ERROR_SUCCESS)
+            {
+            }
+
+            HttpConnection::~HttpConnection()
+            {
+                if (m_connection)
+                {
+                    aws_http_connection_release(m_connection);
+                    m_connection = nullptr;
+                }
+            }
+
+            struct StreamWrapper
```

Comment:
The MQTT API interchangeably names these Wrapper or CallbackData or UserData. Let's pick one and standardize on it. I personally like UserData or CallbackData, since that's what it is.